clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
There were two things I didn't like about the discard approach. First was somewhat stylistic. Having two passes over responses to look at error conditions looks a little smelly. Second, the decision to throw away the data seems arbitrary. Why should we discard good data just because it happened to come in on a request with an unrelated error? And note that we only discard the data from the request that had an out of range error: fetches on other brokers can still succeed, which means there could still be data available in the record set at the same time that an out of range exception is waiting to be thrown. Now that I think about it again, you would probably need the additional call to fetchedRecords for that approach also, so I'm not sure discarding the data really buys you anything.